feat: Added Keptn 0.17 jmeter integration based on job-executor (#172)
keptn-contrib/artifacthub: jmeter/0.3.0/README.md

* `sh.keptn.event.test.triggered`

This can verified in Keptn Bridge -> Project -> Settings -> Integrations -> job-executor-service.

**Example Installation Instructions**
 Please update `JES_VERSION` and `JES_NAMESPACE` in the example below according to your needs.

```bash
TASK_SUBSCRIPTION='sh.keptn.event.test.triggered'
JES_VERSION=0.2.5
JES_NAMESPACE=keptn-jes

helm upgrade --install --create-namespace -n ${JES_NAMESPACE} \
  job-executor-service https://github.com/keptn-contrib/job-executor-service/releases/download/${JES_VERSION}/job-executor-service-${JES_VERSION}.tgz \
  --set remoteControlPlane.autoDetect.enabled="true",remoteControlPlane.topicSubscription="${TASK_SUBSCRIPTION}",remoteControlPlane.api.token="",remoteControlPlane.api.hostname="",remoteControlPlane.api.protocol=""
 ```

**If you have installed jmeter-service, uninstall it**

```bash
helm uninstall jmeter-service -n keptn
```

### Step 2: Build your own JMeter Docker image

> Since there are no official JMeter Docker images, we recommend to build your own (and customize it). Here is a Dockerfile with a basic JMeter installation, feel free to adapt!

Reviewer: ```suggestion
Since there are no official JMeter Docker images, we recommend that you  build and customize your own. Here is a Dockerfile with a basic JMeter installation, feel free to adapt it!
```

	&& curl -L --silent ${JMETER_DOWNLOAD_URL} >  /tmp/dependencies/apache-jmeter-${JMETER_VERSION}.tgz  \
	&& mkdir -p /opt  \
	&& tar -xzf /tmp/dependencies/apache-jmeter-${JMETER_VERSION}.tgz -C /opt  \
	&& rm -rf /tmp/dependencies

# Set global PATH such that "jmeter" command is found
ENV PATH $PATH:$JMETER_BIN

# Entrypoint has same signature as "jmeter" command
COPY entrypoint.sh /

WORKDIR	/keptn/jmeter

RUN chmod +x /entrypoint.sh

ENTRYPOINT ["/entrypoint.sh"]
```

The `entrypoint.sh` file executes JMeter using the passed arguments. You can also uncomment the commands below to enable some debugging steps.

```bash
#!/bin/bash
echo "jmeter args=$@"

echo "START Running Jmeter on `date`"

jmeter $@

echo "END Running Jmeter on `date`"

# Uncomment lines below for debug output
# echo "Test.log"
# cat /keptn/jmeter/test.log

# echo "log.tlf"
# cat /keptn/jmeter/log.tlf
```

Now you will need to build the Docker image and push it to your image registry, e.g., `docker.io/yourorg/jmeter:latest`.

```bash
docker build -t docker.io/yourorg/jmeter:latest .
docker push docker.io/yourorg/jmeter:latest
```

### Step 3: Make sure your project is set up properly with the right tasks.

> While technically not part of the installation instructions, it is worthwhile to mention the coupling between Keptn's shipyard file, and the respective Cloud Event Types configured for Job Executor. 

Reviewer: ```suggestion
While technically not part of the installation instructions, it is worthwhile to mention the coupling between the Keptn shipyard file, and the respective Cloud Event Types configured for Job Executor. 
```

> In example, you should have a monitoring provider (e.g,. Prometheus) configured already for `evaluation` tasks. In addition, your shipyard should contain at least one stages with a `delivery` sequence. Last but not least, the example below also contains a `deployment` step. You can skip this, and just adapt the `JSERVER_URL` in the job-executor config below.

Reviewer: ```suggestion
For example, you should already have a monitoring provider (e.g,. Prometheus) configured for `evaluation` tasks. In addition, your shipyard should contain one or more stages with a `delivery` sequence. Last but not least, the example below also contains a `deployment` step. You can skip this, and just adapt the `JSERVER_URL` in the job-executor config below.
```

Example:
```yaml
apiVersion: "spec.keptn.sh/0.2.2"
kind: "Shipyard"
metadata:
  name: "shipyard-delivery"
spec:
  stages:
    - name: "qa"
      sequences:
        - name: "delivery"
          tasks:
            - name: "deployment"
            - name: "test"
              properties:
                teststrategy: performance
```

### Step 4: Add the Job-Executor configuration file

> The following Job Executor configuration (referred to as job config or `job/config.yaml`) allows you to run jmeter with the `load.jmx` file (e.g., from the [Keptn Carts example](https://github.com/keptn/examples/tree/master/onboarding-carts/jmeter)) using whenever a `test.triggered` with `teststrategy=performance` event is sent by Keptn.

Reviewer: ```suggestion
The following Job Executor configuration (referred to as job config or `job/config.yaml`) allows you to run jmeter with the `load.jmx` file (e.g., from the [Keptn Carts example](https://github.com/keptn/examples/tree/master/onboarding-carts/jmeter)) used whenever a `test.triggered` with `teststrategy=performance` event is sent by Keptn.
```

Add the following content to a file called *jmeter-job-config.yaml* in your current working directory:
```yaml
apiVersion: v2
actions:
  - name: "Run JMeter"
    events:
      - name: "sh.keptn.event.test.triggered"
        jsonpath:
          property: "$.data.test.teststrategy"
          match: "performance"
    tasks:
      - name: "Run jmeter smoke tests"
        files:
          - jmeter/load.jmx
        image: "docker.io/yourorg/jmeter:latest"
        args:
          - '-n'
          - '-t'
          - '/keptn/jmeter/load.jmx'
          - '-JPROTOCOL=http'
          - '-JSERVER_PROTOCOL=http'
          - '-JVUCount=10'
          - '-JLoopCount=10'
          - '-JSERVER_URL=$(KEPTN_SERVICE).$(KEPTN_PROJECT)-$(KEPTN_STAGE).svc.cluster.local'
          - '-j'
          - '/keptn/jmeter/test.log'
          - '-l'
          - '/keptn/jmeter/log.tlf'
```

Add the job-executor configuration file to your Keptn service and stage:

```bash
PROJECT=sockshop
SERVICE=carts

keptn add-resource --project=$PROJECT --service=$SERVICE --stage=staging --resource=jmeter-job-config.yaml --resourceUri=job/config.yaml
```

## Run it

> Now the Job-Executor-Service will execute the JMeter tests whenever you trigger a delivery, e.g., 

Reviewer: ```suggestion
Now the Job-Executor-Service executes the JMeter tests whenever you trigger a delivery, e.g., 
```
